Make test data thread safe

paul-elliott-arm opened the pull request:
## Description

Ensure that the test global data is protected by a mutex.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [ ] **changelog** ~~provided, or~~ not required (test only feature)
- [ ] **backport** ~~done, or~~ not required (new test feature)
- [ ] **tests** ~~provided, or~~ not required (part of current test code)

valeriosetti commented on tests/src/helpers.c:
```diff
@@ -22,6 +26,270 @@ static mbedtls_platform_context platform_ctx;
 
 mbedtls_test_info_t mbedtls_test_info;
```
Now that we have setters/getters for this structure, can't we make this `static`?

valeriosetti commented on tests/src/threading_helpers.c:
```diff
@@ -109,50 +109,70 @@ static void mbedtls_test_mutex_usage_error(mbedtls_threading_mutex_t *mutex,
      * mbedtls_test_mutex_usage_check() will mark it as failed. */
 }
 
+extern mbedtls_threading_mutex_t mbedtls_test_info_mutex;
```
I'm not a huge fan of `extern` variables. What about adding a getter also for this address?
However this is just my opinion, so if you prefer to keep it as it is now, then I'm fine ;)

paul-elliott-arm commented:
> I have a question though since I'm quite new to this topic: is there a way to test these changes? I didn't see any test being added to this PR, but perhaps I'm missing how mutexes are tested (so please apologize if the question looks dumb).

Not a dumb question at all. We have two methods of testing mutexes, one of which that just checks that init->lock->unlock->free all happen in the correct order (However this mutex is excluded from that testing as it would cause a deadlock) and then as part of the CI we run all of the tests under TSAN sanitiser, which is a relatively new thing that will (certainly has) detect race conditions and other potential thread issues.

valeriosetti reviewed (approved):
Thanks for addressing my comments! LGTM :)

Ryan-Everett-arm reviewed (changes requested):
General structure looks good but there is an issue with recording the line1 and line2 variables. I have found a race condition, see my comment for details.

Ryan-Everett-arm commented on tests/include/test/helpers.h:
```diff
@@ -61,20 +62,123 @@ typedef enum {
+/**
+ * \brief           Increment the current test step.
+ */
+void mbedtls_test_increment_step(void);
```
With multiple threads running, the step reported by tests which use this function (currently only dhm tests) will probably be unrelated to where the test failed. To be thread-safe this may need to be removed, and replaced by something which sets `step` in a more meaningful way.

Ryan-Everett-arm commented on tests/suites/host_test.function:
```diff
@@ -715,30 +714,33 @@ int execute_tests(int argc, const char **argv)
-                                    mbedtls_test_info.test);
-                    if (mbedtls_test_info.step != (unsigned long) (-1)) {
+                                    mbedtls_test_get_test());
+                    if (mbedtls_test_get_step() != (unsigned long) (-1)) {
```
Running multiple test functions at once will cause this step variable to be wrongly reported. If one thread fails at step `n`, other threads will keep writing to the step variable and in the final test report the step will likely be wrong.

To keep this step functionality we probably want a way for the first failing thread (the one whose failure is reported) to set `step` in a more permanent way. The issue with this is that calls to `mbedtls_test_set_step` are numerous and the cases are all unique, each "step" corresponds to blocks of code containing multiple checks, and other threads can change `step` at any time; I can't see an easy fix.

Ryan-Everett-arm reviewed (changes requested):
I have been looking into this global `mbedtls_test_info.step` variable - its behaviour is not thread safe, and I'm not sure if this is easily fixable. From what I understand, this field is used in some tests to help a human debugger distinguish at what point of a test a failure occurred. If used, the failure message reads: `test_foo FAILED: step = n; line = m ...`. See my comments for specifics. Sorry for not spotting this in my last review - I wasn't familiar with this variable.

paul-elliott-arm commented:
> I have been looking into this global `mbedtls_test_info.step` variable - its behaviour is not thread safe, and I'm not sure if this is easily fixable. From what I understand, this field is used in some tests to help a human debugger distinguish at what point of a test a failure occurred. If used, the failure message reads: `test_foo FAILED: step = n; line = m ...`. See my comments for specifics. Sorry for not spotting this in my last review - I wasn't familiar with this variable.

You are entirely correct on this, unfortunately. Having spent most of the morning looking at this, I'm not sure there is a way to make this safe that doesn't require reworking of most of the test framework.

But lets keep in mind the scope of what we are attempting to achieve here - we are not attempting to run multiple test suites in seperate threads, we are proposing to run  parts of a single test within threads. I think, therefore it would be enough to comment the step functions heavily to warn that their use within threads will cause undefined results, and obviously not use them in such a manner.

To further stress: using step functions inside threads will not cause crashes or potential race conditions, but the step variable that gets reported in the final test output may well not be what you were expecting. Using them outside of threads will continue to work in the expected manner. Given the amount of work that it would potentially take to fix this, I strongly suggest we backlog a ticket on this, and continue with just the comment for now.

paul-elliott-arm commented:
> >  Given the amount of work that it would potentially take to fix this, I strongly suggest we backlog a ticket on this, and continue with just the comment for now.

Really this only applies to `mbedtls_test_set_step()` as well - with increment, the correct number of increments will be done, just not necessarily in the order one would expect, however the outcome is the same.

Ryan-Everett-arm commented on tests/include/test/helpers.h:
```diff
@@ -215,30 +220,35 @@ void mbedtls_test_fail(const char *test, int line_no, const char *filename);
- *              line number and file name if a test fails. Typically the "step
- *              number" is the index of a for loop but it can be whatever you
- *              want.
+ * \note            Calling this function from a within multiple threads at the
```
Typo: perhaps "It is not recommended for multiple threads to call this function concurrently"

Ryan-Everett-arm reviewed (approved):
LGTM on the basis that the step issue is tracked. There is one typo but it is non-blocking.

valeriosetti reviewed (approved):
LGTM.

Out of my curiosity, one [last comment](https://github.com/Mbed-TLS/mbedtls/pull/8689#issuecomment-1929445116) mentions that 

> Given the amount of work that it would potentially take to fix this, I strongly suggest we backlog a ticket on this, and continue with just the comment for now.

but I don't see any new issue mentioned in the PR's comments after that. Did you already created it?

paul-elliott-arm commented:
Fixed merge conflicts, fixed up the typo in a comment and finally created the issue : https://github.com/Mbed-TLS/mbedtls/issues/8806 (apologies I could have sworn I had done this previously)

valeriosetti reviewed (approved):
LGTM ;)

Ryan-Everett-arm reviewed (approved):
LGTM
